[TECH] Remplacer les plugins good, good-console, good-squeeze qui ont été dépréciées par hapi-pino (Pix-3051).

yahyayouness opened the pull request:
## :unicorn: Problème
Les plugins good, good-console, good-squeeze ont été déprécié depuis 20/12/2020.

## :robot: Solution
Remplacer ces plugins par hapi-pino.
Comme hapi-pino utilise pino, on remplace donc bunyan par pino.

## :rainbow: Remarques
Historique avec good de comment récupérer le métrique route https://github.com/hapijs/good/blob/84de11a53ec38795ffe90e20968acbd17fba1f26/lib/utils.js
Cette PR va garder le comportement par défault du plugin hapi-pino. Une adaptation sera faite coté datadog.
On perd la métrique `route` (à l'image du index.js, ex /users/${id}/foo) qui va être enrichie dans une autre PR.
Pour la métrique event response, il sera mappé dans datadog en se basant sur l'attribut  "msg": "request completed".
J'ai gardé l'event error qui affiche la stacktrace dans un log dédié corrélé par un id de la requête qui en question. On verra par la suite si on veut l'enlever.
Le logger bunyan a été remplacé par le logger pino et la même instance de log a été utilisée par hapi-pino. 
Avec hapi-pino on peut customiser facilement l'objet request remontée dans les logs Datadog, ce qui va nous permettre d'avoir une entrée de logs enrichie avec tous les métriques nécessaires qu'on veut monitorer.

## :100: Pour tester

1. Vérifier en lançant un `npm run test:api:acceptance` que les logs ne sont pas affichés lors des tests.
2. Lancer l'API en locale avec les deux modes dev et production:

Avec un NODE_ENV=developpement npm start , on obtient le log ci-dessous avec l'option prettyPrint pour l'event response:

```
[1630520764051] INFO (7819 on MacBook-Pro-de-yahya.local): request completed
    queryParams: {}
    req: {
      "id": "1630520764034:MacBook-Pro-de-yahya.local:7819:kt1tyl3x:10000",
      "method": "get",
      "url": "/api/admin/users/1",
      "headers": {
        "host": "localhost:3000",
        "user-agent": "curl/7.54.0",
        "accept": "*/*"
      },
      "remoteAddress": "127.0.0.1",
      "remotePort": 50157
    }
    res: {
      "statusCode": 401,
      "headers": {
        "content-type": "application/json; charset=utf-8",
        "vary": "origin",
        "access-control-expose-headers": "WWW-Authenticate,Server-Authorization",
        "cache-control": "no-cache",
        "content-length": 63
      }
    }
    responseTime: 17
```

Avec un NODE_ENV=production npm start , on obtient le json ci-dessous non formaté pour l'event response:

```
{
  "level": 30,
  "time": 1630520449140,
  "pid": 7495,
  "hostname": "MacBook-Pro-de-yahya.local",
  "queryParams": {},
  "req": {
    "id": "1630520449119:MacBook-Pro-de-yahya.local:7495:kt1trufs:10000",
    "method": "get",
    "url": "/api/admin/users/1",
    "headers": {
      "host": "localhost:3000",
      "user-agent": "curl/7.54.0",
      "accept": "*/*"
    },
    "remoteAddress": "127.0.0.1",
    "remotePort": 50128
  },
  "res": {
    "statusCode": 401,
    "headers": {
      "content-type": "application/json; charset=utf-8",
      "vary": "origin",
      "access-control-expose-headers": "WWW-Authenticate,Server-Authorization",
      "cache-control": "no-cache",
      "content-length": 63
    }
  },
  "responseTime": 20,
  "msg": "request completed"
}
```

Et une entrée pour l'event OPS qui n'as pas changé de contrat d'interface par rapport à Good:

```
{
  "level": 30,
  "time": 1630520553900,
  "pid": 7495,
  "hostname": "MacBook-Pro-de-yahya.local",
  "tags": [
    "ops"
  ],
  "data": {
    "host": "MacBook-Pro-de-yahya.local",
    "osload": [
      2.17333984375,
      2.34814453125,
      2.49560546875
    ],
    "osmem": {
      "total": 17179869184,
      "free": 922202112
    },
    "osup": 3872,
    "psup": 113.927457659,
    "psmem": {
      "rss": 126742528,
      "heapTotal": 73125888,
      "heapUsed": 69779264,
      "external": 5699557,
      "arrayBuffers": 4111622
    },
    "pscpu": {
      "user": 3963525,
      "system": 605791
    },
    "psdelay": 0.2529459996148944,
    "requests": {
      "3000": {
        "total": 0,
        "disconnects": 0,
        "statusCodes": {}
      }
    },
    "responseTimes": {
      "3000": {
        "avg": null,
        "max": 0
      }
    },
    "sockets": {
      "http": {
        "total": 0
      },
      "https": {
        "total": 0
      }
    }
  }
}
```
En simulant une API en ERROR:

On obtient la sortie suivante:

```
{
  "level": 50,
  "time": 1630573781837,
  "pid": 12453,
  "hostname": "MacBook-Pro-de-yahya.local",
  "err": {
    "type": "Error",
    "message": "toto",
    "stack": "Error: toto\n    at authenticateUser (/Users/you-octo/Desktop/repositories/pix/api/lib/application/authentication/authentication-controller.js:15:11)\n    at exports.Manager.execute (/Users/you-octo/Desktop/repositories/pix/api/node_modules/@hapi/hapi/lib/toolkit.js:57:29)\n    at Object.internals.handler (/Users/you-octo/Desktop/repositories/pix/api/node_modules/@hapi/hapi/lib/handler.js:46:48)\n    at exports.execute (/Users/you-octo/Desktop/repositories/pix/api/node_modules/@hapi/hapi/lib/handler.js:31:36)\n    at Request._lifecycle (/Users/you-octo/Desktop/repositories/pix/api/node_modules/@hapi/hapi/lib/request.js:372:68)\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)\n    at async Request._execute (/Users/you-octo/Desktop/repositories/pix/api/node_modules/@hapi/hapi/lib/request.js:280:9)",
    "isBoom": true,
    "isServer": true,
    "data": null,
    "output": {
      "statusCode": 500,
      "payload": {
        "statusCode": 500,
        "error": "Internal Server Error",
        "message": "An internal server error occurred"
      },
      "headers": {}
    }
  },
  "msg": "request error"
}
{
  "level": 30,
  "time": 1630573781837,
  "pid": 12453,
  "hostname": "MacBook-Pro-de-yahya.local",
  "queryParams": {},
  "req": {
    "id": "1630573781834:MacBook-Pro-de-yahya.local:12453:kt2pi7mj:10001",
    "method": "post",
    "url": "/api/token",
    "headers": {
      "host": "localhost:3000",
      "user-agent": "curl/7.54.0",
      "accept": "*/*",
      "accept-encoding": "deflate, gzip",
      "content-length": "79",
      "content-type": "application/x-www-form-urlencoded"
    },
    "remoteAddress": "127.0.0.1",
    "remotePort": 51705
  },
  "res": {
    "statusCode": 500,
    "headers": {
      "content-type": "application/json; charset=utf-8",
      "vary": "origin",
      "access-control-expose-headers": "WWW-Authenticate,Server-Authorization",
      "cache-control": "no-cache",
      "content-length": 96
    }
  },
  "responseTime": 3,
  "msg": "request completed"
}
```

yahyayouness committed:
[API] Replace bunyan,good,squeeze,console by pino and hapi-pino for logging

-Replace bunyan by pino And hapi-good, hapi-squeeze, hapi-console by hapi-pino

pix-service commented:
I'm deploying this PR to these urls:

- App (.fr): https://app-pr3380.review.pix.fr
- App (.org): https://app-pr3380.review.pix.org
- Orga: https://orga-pr3380.review.pix.fr
- Certif: https://certif-pr3380.review.pix.fr
- Admin: https://admin-pr3380.review.pix.fr
- API: https://api-pr3380.review.pix.fr/api/

 Please check it out!

octo-topi reviewed (changes requested):
:heart: pour cette PR

Comme vu ensemble, il manque pour être iso-production la reprise des stats systèmes (CPU, mémoire,etc) émis par opsy et récupéré par good

Une implémentation possible ici
https://github.com/GradedJestRisk/js-training/blob/master/node/network/incoming/hapi-logging/code/server.js#L96

yahyayouness commented:
> ❤️ pour cette PR
> 
> Comme vu ensemble, il manque pour être iso-production la reprise des stats systèmes (CPU, mémoire,etc) émis par opsy et récupéré par good
> 
> Une implémentation possible ici
> https://github.com/GradedJestRisk/js-training/blob/master/node/network/incoming/hapi-logging/code/server.js#L96

Merci je suis dessus, oppsy étant aussi déprécié, je cherche une alternative pour y remédier.

octo-topi commented:
> > heart pour cette PR
> > Comme vu ensemble, il manque pour être iso-production la reprise des stats systèmes (CPU, mémoire,etc) émis par opsy et récupéré par good
> > Une implémentation possible ici
> > [https://github.com/GradedJestRisk/js-training/blob/master/node/network/incoming/hapi-logging/code/server.js#L96](https://github.com/GradedJestRisk/js-training/blob/master/node/network/incoming/hapi-logging/code/server.js?rgh-link-date=2021-08-24T10%3A13%3A16Z#L96)
> 
> Merci je suis dessus, oppsy étant aussi déprécié, je cherche une alternative pour y remédier.

J'ai testé ça mais n'ai pas réussi à reproduire les métriques simples existantes de oppsy
https://github.com/radenui/hapi-k8s-health#readme

Je me demande si faire ça depuis Hapi est une bonne idée et s'il ne faudarait pas le faire depuis le conteneur avec prometheus, mais ça devient trop OPS pour moi
https://github.com/GradedJestRisk/js-training/blob/master/node/os/metrics/index.js

Bref, déjà merger cette PR en gardant oppsy ferait baisser le nombre de modules dépréciés, ce serait déjà ça

yahyayouness commented:
> > > heart pour cette PR
> > > Comme vu ensemble, il manque pour être iso-production la reprise des stats systèmes (CPU, mémoire,etc) émis par opsy et récupéré par good
> > > Une implémentation possible ici
> > > [https://github.com/GradedJestRisk/js-training/blob/master/node/network/incoming/hapi-logging/code/server.js#L96](https://github.com/GradedJestRisk/js-training/blob/master/node/network/incoming/hapi-logging/code/server.js?rgh-link-date=2021-08-24T10%3A13%3A16Z#L96)
> > 
> > 
> > Merci je suis dessus, oppsy étant aussi déprécié, je cherche une alternative pour y remédier.
> 
> J'ai testé ça mais n'ai pas réussi à reproduire les métriques simples existantes de oppsy
> https://github.com/radenui/hapi-k8s-health#readme
> 
> Je me demande si faire ça depuis Hapi est une bonne idée et s'il ne faudarait pas le faire depuis le conteneur avec prometheus, mais ça devient trop OPS pour moi
> https://github.com/GradedJestRisk/js-training/blob/master/node/os/metrics/index.js
> 
> Bref, déjà merger cette PR en gardant oppsy ferait baisser le nombre de modules dépréciés, ce serait déjà ça

J'ai remis Oppsy en attendant qu'on trouve mieux. J'avais enlevé du code mort custom qui faisait du prometeus dans l'appli qu'on suspectait à l'origine de memory leak. Ca vaudrait  le coup de revenir dessous et d'avoir un log 

> https://github.com/GradedJestRisk/js-training/blob/master/node/os/metrics/index.js

Yes, j'ai remis oppsy en attendant de trouver mieux. J'avais déjà enlevé du code mort custom qui faisait du prom. On le suspectait à l'origine de mémory leak mais ca vaut le coup de revoir ce qu'on peut garder dans notre contexte.

octo-topi commented:
Proposition: documenter le contrat d'interface OPS avec des tests
Il est implémenté actuellement comme ça
https://github.com/hapijs/good/blob/84de11a53ec38795ffe90e20968acbd17fba1f26/lib/utils.js#L118

octo-topi commented:
Comparaison DEV/PROD existant (avant PR)
```
NODE_ENV=development npm start
210831/152724.556, (1630423644556:OCTO-TOPI:46069:kt084x99:10000) [response] http://OCTO-TOPI:3000: get / {} 404 (14ms)
``` 
+ couleur

PROD 
```
NODE_ENV=production npm start
{"event":"response","timestamp":1630423776764,"id":"1630423776764:OCTO-TOPI:46259:kt087e5b:10002","instance":"http://OCTO-TOPI:3000","method":"get","path":"/","query":{},"responseTime":3,"statusCode":404,"pid":46259,"httpVersion":"1.1","route":"/{p*}","log":[],"source":{"remoteAddress":"127.0.0.1","userAgent":"curl/7.74.0"},"config":{}}
```

Avec PR
```
NODE_ENV=development npm start
[1630423875190] INFO (46605 on OCTO-TOPI): request completed
    req: {
      "id": "1630423875178:OCTO-TOPI:46605:kt08a00f:10000",
      "method": "get",
      "url": "/",
      "headers": {
        "host": "localhost:3000",
        "user-agent": "curl/7.74.0",
        "accept": "*/*"
      },
      "remoteAddress": "127.0.0.1",
      "remotePort": 35950,
      "request_id": "1630423875178:OCTO-TOPI:46605:kt08a00f:10000"
    }
    res: {
      "statusCode": 404,
      "headers": {
        "content-type": "application/json; charset=utf-8",
        "vary": "origin",
        "access-control-expose-headers": "WWW-Authenticate,Server-Authorization",
        "cache-control": "no-cache",
        "content-length": 60
      }
    }
    responseTime: 12
```

PROD 
```
NODE_ENV=production npm start
{"level":30,"time":1630423992730,"pid":46710,"hostname":"OCTO-TOPI","req":{"id":"1630423992712:OCTO-TOPI:46710:kt08cj2s:10000","method":"get","url":"/","headers":{"host":"localhost:3000","user-agent":"curl/7.74.0","accept":"*/*"},"remoteAddress":"127.0.0.1","remotePort":35952,"request_id":"1630423992712:OCTO-TOPI:46710:kt08cj2s:10000"},"req":{"id":"1630423992712:OCTO-TOPI:46710:kt08cj2s:10000","method":"get","url":"/","headers":{"host":"localhost:3000","user-agent":"curl/7.74.0","accept":"*/*"},"remoteAddress":"127.0.0.1","remotePort":35952,"request_id":"1630423992712:OCTO-TOPI:46710:kt08cj2s:10000"},"res":{"statusCode":404,"headers":{"content-type":"application/json; charset=utf-8","vary":"origin","access-control-expose-headers":"WWW-Authenticate,Server-Authorization","cache-control":"no-cache","content-length":60}},"responseTime":18,"msg":"request completed"}
```

yahyayouness committed:
[API] Add hapi-oppsy dependency and config for event ops

yahyayouness committed:
[API] Remove duplicate log stacktrace error after use hapi-pino

After replacing hapi-good by hapi-pino, the log contain by default error event stacktrace so we don't need anymore to log it

yahyayouness committed:
[API](SR) move check prod env in setting configuration
